### ISSUE TITLE
Automatic conversion from short2 in vertex data to float2 in shader fails under HLSL5

### ISSUE THREAD

**Issue by mattiasljungstrom** [floooh/sokol-tools#6]
This might be a limitation in HLSL/D11, I'm not sure. Figured I'll report it. There's a workaround, see below.

I have the following setup for vertex data:
        float x, y;
        uint8_t[4] c;
        uint16_t u, v;

The pipeline is setup as:
        attrs[0].format = SG_VERTEXFORMAT_FLOAT2;
        attrs[1].format = SG_VERTEXFORMAT_UBYTE4N;
        attrs[2].format = SG_VERTEXFORMAT_SHORT2;

In the shader code I use this setup, which works under OpenGL and Metal:
layout(location=0) in vec2 a_position;
layout(location=1) in vec4 a_color;
layout(location=2) in vec2 a_uv;

Note that color is correctly converted from uint8 to vec4 floats, but it fails for the uv data.

Workaround is to define a_uv as ivec2, like this:
layout(location=2) in ivec2 a_uv;

And then in shader code do: vec2(a_uv)

**Comment by floooh** [comment]
I need to check when I'm near a Windows machine. Looks like there are no sokol- or oryol-samples which use such a conversion, but we're using packed 16-bit integers for texture coordinates in Drakensang with D3D9, not sure if the rules have changed for D3D10/11

The ivec2 workaround most likely won't work in GLES2/WebGL, in case that's important for you :)

**Comment by floooh** [comment]
Ok I think I know what the issue is:

D3D11 doesn't convert from integer to float unless it's a normalized format (so UBYTE4N to vec4 and SHORT2N to vec2 is ok, but UBYTE4 to vec4 or SHORT2 to vec2 isn't)...

So a packed vertex format which is portable across all backend APIs would need to use only the normalized integer formats (with the leading 'N'). Those can be expanded in the vertex-shader by multiplying with 127.0, 255.0, 32767.0 or 65535.0).

I'll modify the texcube-sapp sample to use UBYTE4N for colors and SHORT2N for texture coords.

The details are in the recently released D3D11 spec: https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#DataConversion

**Comment by floooh** [comment]
...I'll explain that in the sokol_gfx.h documentation comment block and the texcube-sample as well...

**Comment by mattiasljungstrom** [comment]
Ah, yes of course. I tried SHORT2N but forgot to multiply with 32767.0 to get the correct result. Thanks for looking into it!

**Comment by nmr8acme** [comment]
Maybe ignorant question but to convert from SHORT2N to vec2 exactly as GL/Metal would, don't you need to do:

```
vec2 input;
vec2 decoded_input = (input + 1.0) * 65535.0 - 32767.0;
```

... instead of ...

```
vec2 input;
vec2 decoded_input = input * 32767.0;
```

... because the range of int16_t is [-32767, 32768]? But the range of the simpler decoding is [-32767, 32767], which appears to be off by one?

Lots of things I am unsure of here (including precision and rounding,) so take this with a grain of salt.

(My motivation is that I have some code that uses the LSB in a SHORT2 attribute to encode a flag and I want to make sure when I convert it to SHORT2N it behaves exactly the same WRT the LSB)

**Comment by nmr8acme** [comment]
Okay, the answer to my question is "no", you don't need the more complicated conversion because the mapping from SHORT2N to float is not what I expected. See https://registry.khronos.org/OpenGL/specs/gl/glspec46.core.pdf section 2.3.5.1 and this Mesa ticket (https://mesa-dev.freedesktop.narkive.com/d2rY38Zo/snorm-conversion-equation-woes) for some background.

(Also I derped on my int16_t max/min values, the max is 32767 and the min is -32768, not ~vice versa as I had written.)

Basically when mapping int16_t values to float values, -32767 is mapped to -1.0 and 32767 is mapped to 1.0, and -32768 is a redundant encoding of -1.0. This is so that 0.0 can be exactly encoded.
